shared/src/main/scala/coop/rchain/dag/DagOps.scala

```diff
@@ -0,0 +1,158 @@
+                             acc.asRight[(Seq[A], Seq[A])].pure[F]
+                           case blocks =>
+                             for {
+                               nextBlocks <- blocks
+                                              .traverse(
+                                                b =>
+                                                  dag
+                                                    .parents(b)
+                                                    .map(_.getOrElse(Set.empty))
+                                              )
+                                              .map(_.flatten)
+                               // stop when base block is met
+                               next = if (nextBlocks.contains(base)) Seq.empty else nextBlocks
+                             } yield (acc ++ next, next).asLeft[Seq[A]]
+                         }
+                     }
+      // blocks from tipSet connected to base
+      branchSet <- (Seq.empty[A], Seq(base))
+                    .tailRecM[F, Seq[A]] {
+                      case (acc, blocksToProceed) =>
+                        blocksToProceed.toList match {
+                          case Nil =>
+                            acc.asRight[(Seq[A], Seq[A])].pure[F]
+                          case blocks =>
+                            for {
+                              children <- blocks
+                                           .traverse(
+                                             b =>
+                                               dag
+                                                 .children(b)
+                                                 .map(_.getOrElse(Seq.empty))
+                                           )
+                                           .map(_.flatten)
+                              childrenFromTipSet = children.filter(tipAncestors.contains)
+                            } yield (acc ++ childrenFromTipSet, childrenFromTipSet).asLeft[Seq[A]]
+                        }
+                    }
+    } yield (tip, branchSet)
+
+  /**
+    * @return seq of branches, sorted by size starting with the biggest one
+    */
+  def computeSortedBranches[F[_]: Concurrent, A](
+      tips: Seq[A],
+      base: A,
+      dag: DagReader[F, A]
+  )(implicit ord: Ordering[A]): F[Seq[(A, Seq[A])]] =
+    for {
+      // compute branches in parallel, sort according to size primarily and tip secondary, in reverse order
+      branches <- fs2.Stream
+                   .emits(tips.map { top =>
+                     fs2.Stream.eval(computeBranch(top, base, dag))
+                   })
+                   .parJoinUnbounded
+                   .compile
+                   .toList
+                   .map(_.sortBy(b => (b._2.size, b._1)))
+      // make branches non intersecting
+      (_, nonIntersectingBranches) = branches.reverse
+        .foldLeft(
+          // initial list of branches and result list of non intersecting branches
```

Comment:
What is intersacting branches?
B4 and B5 like below?
```
    B4       B5
   /    \    /    \
  B1    B2     B3
     \      |     /
        base
```

Reply:
B4  B1  B2  base
B5  B3  B2  base
these are intersecting
to merge we deduplicate them so have
B4 B1 B2 base
B5 B3